test/management_tests/management_test.py

```diff
@@ -223,3 +223,62 @@ async def test_broadcast_message():
 
     with pytest.raises(TypeError):
         await xknx.management.broadcast(connect)
+
+
+@pytest.mark.parametrize("rate_limit", [0, 1])
+async def test_p2p_rate_limit(time_travel, rate_limit):
+    """Test rate limit for P2P management connections."""
+    xknx = XKNX()
+    xknx.cemi_handler = AsyncMock()
+    ia = IndividualAddress("4.0.1")
+
+    def send_responses(index):
+        ack = Telegram(
+            source_address=ia,
+            destination_address=IndividualAddress(0),
+            direction=TelegramDirection.INCOMING,
+            tpci=tpci.TAck(index),
+        )
+        device_desc_resp = Telegram(
+            source_address=ia,
+            destination_address=IndividualAddress(0),
+            direction=TelegramDirection.INCOMING,
+            tpci=tpci.TDataConnected(index),
+            payload=apci.DeviceDescriptorResponse(),
+        )
+
+        xknx.management.process(ack)
+        xknx.management.process(device_desc_resp)
+
+    conn = await xknx.management.connect(ia, rate_limit)
+
+    # create task and request data
+    task = asyncio.create_task(
+        conn.request(
+            payload=apci.DeviceDescriptorRead(descriptor=0),
+            expected=apci.DeviceDescriptorResponse,
+        )
+    )
+
+    await asyncio.sleep(0)
+    send_responses(0)
+
+    await task
+
+    # create second task
+    task = asyncio.create_task(
+        conn.request(
+            payload=apci.DeviceDescriptorRead(descriptor=0),
+            expected=apci.DeviceDescriptorResponse,
+        )
+    )
+    await asyncio.sleep(0)
+    send_responses(1)
+
+    if rate_limit > 0:
+        # this should raise a timeout, since the data was not yet sent
+        with pytest.raises(ManagementConnectionTimeout):
+            await task
```

Comment:
I haven't used management for a while, so I'm not really following this.

What I see here:
L270 creates a task to request a response from a device. It isn't sent because still in the rate limit of the previous task - so it would theoretically wait 1 second before sending.

L276 `send_responses(1)` sends responses for a request it hasn't even received yet.
Since not yet waiting for an ACK, it is discarded (`process()`).
Same for the `DeviceDescriptorResponse` since not yet waiting for that sequence_number. (I guess `Management.process()` sends an ACK for that).

L281 awaiting the response we previously discarded - we block for a second waiting for nothing.

...
If this is correct, I guess we should at least `time_travel` in the `pytest.raises` block. And probably don't send the response beforehand, but look for the sent request - before and after the `time_travel`.

Reply:
That's an alternative. The idea of the test was to do exactly the same sequence with and without the rate limit option. The new version checks for the sent messages.